src/main/telemetry/msp_shared.c

```diff
@@ -24,230 +24,294 @@
 
 #include "platform.h"
 
+//hack for intellisense. Don't push these changes
+#ifndef USE_MSP_OVER_TELEMETRY
+#define USE_MSP_OVER_TELEMETRY
+#endif
+#include "io/vtx_smartaudio.h"
+
 #if defined(USE_MSP_OVER_TELEMETRY)
 
 #include "build/build_config.h"
 
 #include "common/utils.h"
+#include "common/crc.h"
+#include "common/streambuf.h"
 
 #include "msp/msp.h"
 #include "msp/msp_protocol.h"
+#include "msp/msp_serial.h"
 
 #include "telemetry/crsf.h"
 #include "telemetry/msp_shared.h"
 #include "telemetry/smartport.h"
 
-#define TELEMETRY_MSP_VERSION    1
-#define TELEMETRY_MSP_VER_SHIFT  5
-#define TELEMETRY_MSP_VER_MASK   (0x7 << TELEMETRY_MSP_VER_SHIFT)
-#define TELEMETRY_MSP_ERROR_FLAG (1 << 5)
-#define TELEMETRY_MSP_START_FLAG (1 << 4)
-#define TELEMETRY_MSP_SEQ_MASK   0x0F
+
+/*
+---------------------------------------------------------------
+How MSP frames are sent over CRSF:
+CRSF frame types: 0x7A for MSP requests, 0x7B for responses.
+CRSF extended header frames are used. i.e., Destination and origin addresses added after CRSF type.
+CRSF frame structure:
+<sync/address><length><type><destination><origing><status><MSP_body><CRC>
+Status byte consists of three parts:
+    bits 0-3 represent the sequence number of the MSP frame,
+    bit 4 checks if current MSP chunk is the beginning of a new frame (1 if true),
+    bits 5-6 represent the version number of MSP protocol (MSPv1 or MSPv2)
+    bit 7 represents Error (1 if there was an error)
+MSP_body is unmodified MSP frame without header ($ + M|X + <|>|!) and CRC.
+MSP might be MSPv1 or MSPv2 or MSPv1_Jumbo.
+
+MSP_body might be sent in chunks.
+First (or only) chunk must always set "begin of frame" bit (#4) of status byte.
+Each next chunk must have increased frame number in status byte.
+Size of chunk is recovered from size of CRSF frame. Although last / only CRSF frame might have size bigger than needed for MSP-body. Extra bytes must be ignored. So, the real size of MSP-body must be parsed from the MSP-body itself.
+CRSF frames might be any size until maximum of 64 bytes for a CRSF frame. So, maximum chunk size is 57 bytes. Although, MSP-body might be sent in shorter chunks.
```

Comment:
These are long lines. Can they be split up to make it easier to read without scrolling? 
Other than that the documentation here is good. 👍